data-hibernate-jpa/src/main/java/io/micronaut/data/hibernate/operations/HibernateJpaOperations.java

```diff
@@ -387,10 +389,17 @@ public <T> Iterable<T> persistAll(@NonNull InsertBatchOperation<T> operation) {
     }
 
     private void flushIfNecessary(EntityManager entityManager, AnnotationMetadata annotationMetadata) {
+        flushIfNecessary(entityManager, annotationMetadata, false);
+    }
+
+    private void flushIfNecessary(EntityManager entityManager, AnnotationMetadata annotationMetadata, boolean clear) {
         if (annotationMetadata.hasAnnotation(QueryHint.class)) {
             FlushModeType flushModeType = getFlushModeType(annotationMetadata);
             if (flushModeType == FlushModeType.AUTO) {
                 entityManager.flush();
+                if (clear) {
+                    entityManager.clear();
```

Comment:
Must clear session in case when executeUpdate is used to remove current items in session, or else these will still be retrieved.